CMakeLists.txt

```diff
@@ -1,284 +1,41 @@
-  endif() # USE_NIX
-endif() # APPLE
-
-find_package(PkgConfig REQUIRED)
-pkg_check_modules(FFI REQUIRED libffi)
-if(USE_NIX)
-include_directories(AFTER ${FFI_INCLUDE_DIRS})
-link_directories(AFTER ${FFI_LIBRARY_DIRS})
-else() # USE_NIX
-include_directories(AFTER SYSTEM ${FFI_INCLUDE_DIRS})
-link_directories(AFTER ${FFI_LIBRARY_DIRS})
-endif() # USE_NIX
-
-if ((CMAKE_BUILD_TYPE STREQUAL "Release" OR CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
-     AND K_LLVM_BACKEND_LTO)
-set(LLVM_KOMPILE_LTO "lto")
-else()
-set(LLVM_KOMPILE_LTO "nolto")
-endif()
-
-set(CMAKE_AR "${CMAKE_C_COMPILER_AR}")
-set(CMAKE_RANLIB "${CMAKE_C_COMPILER_RANLIB}")
-
-find_package(LLVM 15 QUIET CONFIG)
 
-if (NOT LLVM_FOUND)
-  find_package(LLVM 14 QUIET CONFIG)
-endif()
-
-if (NOT LLVM_FOUND)
-  find_package(LLVM 13 QUIET CONFIG)
-endif()
-
-if (NOT LLVM_FOUND)
-  find_package(LLVM 12 QUIET CONFIG)
-endif()
+include(CTest)
 
-if (NOT LLVM_FOUND)
-  find_package(LLVM 11.1 QUIET CONFIG)
-endif()
+# Needs to be included first so that pkg-config paths are set up when looking
+# for dependencies.
+include(FixHomebrew)
 
-if (NOT LLVM_FOUND)
-  find_package(LLVM 11 REQUIRED CONFIG)
-endif()
+add_subdirectory(deps)
+include(FindDependencies)
 
-message(STATUS "Found LLVM ${LLVM_PACKAGE_VERSION}")
-if (${LLVM_PACKAGE_VERSION} VERSION_LESS 11)
-  message(FATAL_ERROR "LLVM 11 or newer is required")
-endif()
+include(KLLVMCompilerFlags)
 
 include(SetTriple)
 
-set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
-
-find_package(ZLIB REQUIRED)
-find_package(FLEX REQUIRED)
-find_package(GMP REQUIRED)
-find_package(Boost REQUIRED COMPONENTS unit_test_framework)
-
-include_directories(${LLVM_INCLUDE_DIRS})
-include_directories (${PROJECT_SOURCE_DIR}/include)
-
-macro(kllvm_add_tool name)
-  add_executable(${name} ${ARGN})
-
-  llvm_config(${name}
-    USE_SHARED true
-    core irreader
-  )
-
-  if(APPLE)
-    if(NOT USE_NIX)
-      target_link_libraries(${name} PUBLIC "-ljemalloc" "-Wl,-rpath ${BREW_PREFIX}/lib" "-ldl")
-    else()
-      target_link_libraries(${name} PUBLIC "-ljemalloc" "-ldl")
-    endif() # NOT USE_NIX
-  else()
-    if(NOT USE_NIX)
-      target_link_libraries(${name} PUBLIC "-ljemalloc" "-Wl,-rpath /usr/local/lib" "-ldl")
-    else()
-      target_link_libraries(${name} PUBLIC "-ljemalloc" "-ldl")
-    endif() # NOT USE_NIX
-  endif() # APPLE
-endmacro(kllvm_add_tool)
-
-configure_file(runtime/main/main.ll ${CMAKE_CURRENT_BINARY_DIR} @ONLY)
-install(
-  FILES ${CMAKE_CURRENT_BINARY_DIR}/main.ll runtime/main/search.cpp
-  DESTINATION lib/kllvm/llvm/main
-)
-
-find_program(OPT NAMES opt-14 opt-13 opt-12 opt-11 opt-10 opt)
-find_program(LLC NAMES llc-14 llc-13 llc-12 llc-11 llc-10 llc)
-if(${OPT} STREQUAL "OPT-NOTFOUND")
-  message(FATAL_ERROR "Could not find an opt binary. Is llvm installed on your PATH?")
-endif()
-if(${LLC} STREQUAL "OPT-NOTFOUND")
-  message(FATAL_ERROR "Could not find an llvm binary. Is llvm installed on your PATH?")
-endif()
```

Comment:
This is the primary change that motivated my cleanup; this code has been extracted to `FindLLVM` and the logic changed so that we look for `opt` and `llc` in the LLVM binary tools directory for the package we found with CMake, rather than punting to a versioned search.

There is also a small refactoring to simplify the boolean checks.